cpp/src/io/avro/avro.cpp

```diff
@@ -92,23 +94,38 @@ bool container::parse(file_metadata* md, size_t max_num_rows, size_t first_row)
       }
     }
   }
+  // Save the first sync markers in the metadata; we compare them to other
+  // sync markers that should be present at the end of a block.  If they
+  // differ, the data should be interpreted as corrupted.
   md->sync_marker[0] = get_raw<uint64_t>();
   md->sync_marker[1] = get_raw<uint64_t>();
 
   md->metadata_size  = m_cur - m_base;
-  md->skip_rows      = 0;
+  md->skip_rows      = first_row;
```

Comment:
I don't understand this fix. At a glance, the old logic makes sense to me: iterate over blocks and reduce the number of rows to skip until we don't need to skip all rows in the current block. Record the number of rows to skip in the first block that is included in block_list (which only has block that we at least partially read).
What was incorrect in this logic?

Reply:
I'm going to write some multi-block tests and come back to this (very reasonable) question.  I am re-reading the code and I'm inclined to agree -- perhaps the only fix that's needed is this bit that was at the end of the while loop:

```cpp

  if ((max_num_rows <= 0) || (max_num_rows > total_object_count)) {
    md->num_rows = total_object_count;
  } else {
    md->num_rows = max_num_rows;
  }
```

Reply:
Addressed in the most recent commit.